packages/core/src/enhance/impl.js

```diff
@@ -1,39 +1,59 @@
 'use strict'
 
 import R from 'ramda'
-import { memoize, time } from '../impl/util'
+import { memoize } from '../impl/util'
 import * as data from '../data'
 import * as zip from '../zip'
 
-export function enhancer(config) {
-  const { registry, elementZipper } = config
+export function extract(config) {
+  const { registry, elementZipper, minNumOfArgs, maxNumOfArgs } = config
 
-  const enhancersForKind = memoize(kind => {
-    const enhancers = registry.get(kind)
-    return enhancers.isEmpty() ? null : enhancers
-  })
+  const enhancersForKind = memoize(kind => registry.get(kind))
 
-  async function enhance(loc, context) {
-    const { elementZipper } = context
-
-    const kind = data.flow(loc, zip.value, data.kindOf)
-    const enhancers = enhancersForKind(kind)
-    if (enhancers != null) {
+  async function _extract(loc, context) {
+    const kind = data.flow(
+      loc,
+      zip.value,
+      el => el.update('kind', k => k.filterNot(R.equals('__loading'))),
```

Comment:
Why this here? Something of kind `__loading, X` is not `X` and should not be treated like that. As little as possible surprises, please

Reply:
This badly named `extract` can be executed in parallel with the read. At that point, the kind of the element might be something like `['__loading', 'document', 'page']`. An enhancer registered for `['document', 'page']` will not work without this line of code.